### DISCUSSION TITLE
How bulk operations work

### DISCUSSION THREAD

**Discussion by MapInetum** [Mimetis/Dotmim.Sync#883] [Q&A] [answered]
Hi all.
I don't understand how DMS works to determine which stored procedure (bulk or not bulk) to use when updating N records.
I use sql server change tracking.
For example:
I update 5 fields and according to the log, I see q does a bulkupdate first and then the 5 updates individually.
Al I am not understanding well...

![image](https://user-images.githubusercontent.com/79869950/208688028-a39279f4-b8bd-49a8-b4f1-7ff2bf490f17.png)

**Comment by Mimetis** [comment]
Hi @maprado-inetum 

When a batch is applied, sometimes, we can have an error during the `bulkupdate` step (for any reason, `Primary Key Violation`, `Foreign Key Error`, `Null Exception` and so on ...)

Since it's a "All the Batch or Nothing", we don't know which row causes the error...
That's why if we have an error on a batch update, DMS tries again this batch using a _per line update_.

During the per line update, we will know which row has an error preventing the bulk update to happen.

The thing I don't understand in your screenshot is that it misses a `LOG INF` saying that `DMS` is fallback to _per line update_ as we can see in the source code:

https://github.com/Mimetis/Dotmim.Sync/blob/490bb41189f638ed7dfce82136a2f0ce235e84b7/Projects/Dotmim.Sync.Core/Orchestrators/ApplyChanges/BaseOrchestrator.ApplyChanges.cs#L425

You seem to have a situation where there is no error from the `bulkupdate`, but **DMS** still tries to make a per line update ....

1. Which `DMS` version are you using? (I'm wondering if a previous version may not have the `LOG INF` in the source code)
   - If you are not using the last version, can you try with the latest version ?
3. Are you able to easily reproduce the bug?
    - If yes, are you able to create a sample repro, and share it with me?

Let me know

**Reply by MapInetum** [reply]
Hi Sébastien.
I'm utilizing v0.9.7 of DMS.
I prepare this example for you but I'm not able to reproduce it, sorry.
The first sync example is in Production enviroment and is not possible share with anyone.

[BulkUpdates.Example.zip](https://github.com/Mimetis/Dotmim.Sync/files/10114118/BulkUpdates.Example.zip)

**Reply by Mimetis** [reply]
Ok, let me know if the problems persists.
If yes, we need a way to reproduce it

**Comment by MapInetum** [comment]
The problem is reproduced in production again....
But I don't know why.

![image](https://user-images.githubusercontent.com/79869950/204818380-cde72969-4077-47ee-95d0-34027ad6836d.png)

**Reply by Mimetis** [reply]
Did it made **667** single insert ?

**Reply by MapInetum** [reply]
Yes... And it's very slowly.
I supose the problem is with CHANGE TRACKING. What do you think?

**Reply by Mimetis** [reply]
I don't know
I've published a new beta version `0.9.8` that is basically the same as actual production `0.9.7`, but with more details in the logs and errors.
You should try it and see if you have more information.
As the messages returned from errors are different (and error type as well) be sure it won't break anything for you and test it

**Reply by Mimetis** [reply]
When DMS fallbacks to single per line row, we will have the error message in the logs

**Reply by MapInetum** [reply]
Nothing, nothing.
I found.

![image](https://user-images.githubusercontent.com/79869950/204843085-d2dbc196-92da-4047-b823-1b5287e2a32e.png)

**Reply by Mimetis** [reply]
use the 0.9.8 beta u see in the screenshot ;)

**Comment by MapInetum** [comment]
I update my project with DMS v0.9.8 and when the application runs it breaks with this exception:

![image](https://user-images.githubusercontent.com/79869950/204918284-8b210e80-98b2-4d9f-9fa7-b74f72292881.png)

These are the nugets updates that my project needs to update to DMS v0.9.8:

![image](https://user-images.githubusercontent.com/79869950/204918449-466f254c-4c46-4bf7-8661-e3001d0accda.png)

My project is NET Core 5 and with DMS v0.9.7 run 'ok'.

What change?

**Reply by Mimetis** [reply]
I dealt with the same kind of problems recently.
I don't know why, but it was resolved by adding `TrustServerCertificate=True;` to the connection string.

**Reply by Mimetis** [reply]
I think it comes from here : https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/breaking-changes#encrypt-true (not sure, just an idea)

**Reply by Mimetis** [reply]
Looking at your stacktrace, I see you are using EF, so I guess this is the problem you encountered.

**Reply by Mimetis** [reply]
If trust server certificate does not work, remove `TrustServerCertificate=True` and replace with `Encrypt=False` in your connection string

**Reply by vunder** [reply]
The problem is not exactly in EF. It just triggered it. There is "end of life" for "old" ssl protocols. TLS 1.0 and 1.1 are deprecated. A lot of services were already updated. EF is not the exception. MSSQL 2014 cannot use TLS 1.2 that is why you cannot connect to it.
@maprado-inetum you need either update you database service or, as suggested @Mimetis , use security weak connection by declaring `TrustServerCertificate=True` in connection string

**Comment by MapInetum** [comment]
Great!
I think with this version bulkupdates works fine.

![image](https://user-images.githubusercontent.com/79869950/204999025-c705807b-1cf4-4cb7-9f32-87a6a29c93d4.png)

**Reply by Mimetis** [reply]
Ok, let me know if the last issue appears again.
I guess logs will be more verbose with more info

**Comment by MapInetum** [comment]
Hello again.
I try to explain what I have seen with a sicronization made in a production environment this morning.
The first synchronization performed with a high number of records (4M) has finished successfully using bulkupdates of about 10,000 in 10,000 records.

![image](https://user-images.githubusercontent.com/79869950/205032708-08542d72-a9f9-4697-aab2-7a8a0357409b.png)


However, the next synchronization of about 148 records, the problem has been reproduced again as I show below.
The only difference between the first synchronization and the second one, is that in this second one other clients were also synchronizing in parallel at the same time against the same server.

![image](https://user-images.githubusercontent.com/79869950/205033027-e7dcfc7d-0751-49da-9f86-907ea31415a2.png)


Could it be that the server decides to switch from bulkpudate to singleupdate when there are many interlocks?

**Reply by Mimetis** [reply]
Maybe.
Problem here is that you still don't have the error message.
Are you sure you 've migrated all the clients / server to the last beta version `0.9.8` ?

**Reply by MapInetum** [reply]
No, I don't use DMS v0.9.8 on all clients, I thought it was only necessary to use it on one client to reproduce the error.
Getting it to or all production clients is going to take me some time.
I think the best thing to do is to prepare a development environment and try to reproduce it again.
It's going to take me time to do it...

**Reply by Mimetis** [reply]
The error log is coming from which side ? a client (with version 0.9.8) or the server (which version ?)

**Reply by Mimetis** [reply]
I'm not sure, but I guess the screenshot is from the server side, right ?
If so, you don't need to upgrade clients, as the problem is coming from the server, right ?

If yes, you just need to upgrade the server as the version `0.9.8` is a minor version, fully compatible with `0.9.7`

**Comment by MapInetum** [comment]
Hello everyone.

I would like to take up this problem again, as it happens to me continuously in a productive environment.

Basically, what happens to me is what I showed previously. 
When several records in the table are updated and the synchronization is launched, I see in the logs that in the target database is executed first the mass update stored procedure "TableName_ScopeName_bulkupdate" but then (without any error or exception), is executed the unit update stored procedure "TableName_ScopeName_update".

This causes the synchronization of thousands of records to take forever and leads to crashes with the other synchronization consoles.

Has this happened to anyone?
Thanks!

**Reply by Mimetis** [reply]
Bulk Update tries to update a batch of rows in one single instruction, using TVP and stored procedures.

If an error occurs during the batch operation, SQL Server will not returns the error row that causes the error.
That's why, if an error is making the batch failing, DMS gets all the rows from the batch, and tries to apply them 1 by 1 to catch the error.

The code is here (see fallback line ~~420):

https://github.com/Mimetis/Dotmim.Sync/blob/3d8a2fa7a0de84d0fe0e8906eb067ff48d4f3198/Projects/Dotmim.Sync.Core/Orchestrators/ApplyChanges/BaseOrchestrator.ApplyChanges.cs#L384-L467

If you are using the last pre release version, you may have the error in the logs

**Comment by MapInetum** [comment]
I understand that I should see a log trace like the one on line 425.

In my productive environment I don't see that trace and we have configured the `SyncProgressLevel = Sql`

The version of Dotmim.Sync is 0.9.8-beta-0994.

Yesterday log for example:

![image](https://user-images.githubusercontent.com/79869950/213469387-18c58e13-3750-4dad-a85c-d8da289bbe9e.png)

**Reply by Mimetis** [reply]
Is it log from the server side or from the client side ?
What is the default strategy for resolving conflict ? ServerWins or ClientWins ?
Do you have conflict resolution count after this huge update batch ?

**Comment by MapInetum** [comment]
These log traces are server-side.

I think at the time you proceed to execute the stored procedures on the server.

We use a ClientWins strategy.

And no, it does not produce a conflict count.

**Reply by Mimetis** [reply]
Interesting ...
I don't see, in the code, where the Update call can be made without a log..
That's strange and weird ....

Are you able to reproduce the bug easily or is it nearly impossible for you to provide a sample ?

**Comment by MapInetum** [comment]
I could provide you with a test scenario, but it is impossible for me to reproduce it.

In the test/development scenario, I work with local DBs and I think this hides what might be happening...

[Dotmim.Sync - #935 Q&A.zip](https://github.com/Mimetis/Dotmim.Sync/files/10457874/Dotmim.Sync.-.935.Q.A.zip)

**Comment by MapInetum** [comment]
One question.
How do the locks work between the different transactions when executing the stored procedures in this case on the server?

**Reply by Mimetis** [reply]
depends if you have the options `DisableConstraintsOnApplyChanges` or not

If `DisableConstraintsOnApplyChanges==true` then the entire table is locked during the operation (because we are disabling the constraints for all rows)

If `DisableConstraintsOnApplyChanges==false` then only the rows updated / deleted are locked (well, the leaf pages in reality)

**Comment by MapInetum** [comment]
Good night!

@Mimetis I think that you can reproduce my problem with this scenario.

All necesary is in the ZIP file. View README first.

Thanks for all!

https://we.tl/t-gTdSAOVKLl

**Reply by Mimetis** [reply]
can you share the console source code ?

**Reply by Mimetis** [reply]
I made the test and did not notice anything strange...
The logs you are outputting on the console makes everything very slow, but at least it worked.
Will take a look at the source code if you can share it with me.

![image](https://user-images.githubusercontent.com/4592555/213825570-38b971a3-3c6d-4bca-a124-ca9a046d39ef.png)

**Reply by MapInetum** [reply]
Good morning.

This is the source code of the console.

[Dotmim.Sync.Console(Source).zip](https://github.com/Mimetis/Dotmim.Sync/files/10471786/Dotmim.Sync.Console.Source.zip)

Making massive inserts of 100000 records in the 10 clients at the same time.

You can see how from executing the procedure of massive inserts/updates, it uses the unitary one.

![image](https://user-images.githubusercontent.com/79869950/213859055-1dca5236-1f2e-4ac7-96cc-e8c8a8f0bf84.png)

View of SQL Server instance activity monitor

![image](https://user-images.githubusercontent.com/79869950/213861169-e1c3b161-4bfe-40e1-867f-0b420050a27d.png)

**Reply by Mimetis** [reply]
I'm working on it.
I found what the problems are:
- **Connection Timeout** : Seems connection timeout was not applied on the bulkupdate command. It's fixed (but it was not your problem)
![timeout](https://user-images.githubusercontent.com/4592555/213933086-c04193f0-5b85-4e17-8732-3eb8d454cff4.png)

- **Deadlocks** : Deadlocks makes the batch fail
![deadlock](https://user-images.githubusercontent.com/4592555/213933096-440108cc-6d0a-496e-a086-61f13b6988af.png)

The problem with deadlocks is that it really depends on a lot of factors.
I'm working on resolving the issue, but I'm facing others strange behaviors.

What I can suggest so far:
- Makes smaller bulk insert batch: **50 000** is too much and it will lock too much leaf pages. Use between **5 000** and **10 000**
- Use Full Transaction on Server side (not batch)
- Use Snapshot isolation on server side (not from DMS but from the databse itself):

``` sql
Alter database [{dbName}] SET ALLOW_SNAPSHOT_ISOLATION ON;  
Alter database [{dbName}] SET READ_COMMITTED_SNAPSHOT ON;
```
Try to limit logs to INFORMATION as the other one will grab the query (and then will slow dramatically the time to apply a batch)

I'm still working on it but will not be available this week.
Will continue to work on it early February

**Reply by MapInetum** [reply]
Thank you very much, Sébastien.

I will follow your advice to minimize problems in the production environment.

We will continue in February ;)

**Comment by MapInetum** [comment]
Hello.

I minimized problems with this configuration:

- In DMS, when configuring the server provider, isolation level is ReadUncommitted
- I used per batch transactions on the serve side.

I used this configuration because in production enviroment it's not possible use snapshot isolation.

with this the production synchronizations are stable but very slow.

My biggest problem is still the unit inserts.

I need a solution for DMS to use bulk insertion (stored procedures 'bulkupdate' and 'bulkdelete').

Regards.

**Reply by Mimetis** [reply]
Hey @maprado-inetum 
I'm currently working on a solution where transient errors are handled correctly (like deadlocks for instance)
In that case, you will have the information in the logs, and even an interceptor.
The transaction will be rerun a few seconds later (5 times) before raising an error.

I need a little bit more time to stabilize and test everything, but I guess I should have something available until the end of January.
Will update you as soon as I have something to share.

**Reply by Mimetis** [reply]
hey @maprado-inetum 
I've release a new pre-release version [0.9.8-beta-1375](https://www.nuget.org/packages/Dotmim.Sync.Core/0.9.8-beta-1375) that's handle transient errors (see PR #949)

You should see less fallback to _line per line_ inserts / updates due to timeout or deadlocks.

In my opinion, you still should to:
- Makes smaller bulk insert batch: 50 000 is too much and it will lock too much leaf pages. Use between 5 000 and 10 000
- Batch mode or Full transaction on server side is ok now.
- Try to limit logs to INFORMATION

In logs you will have now more information when deadlocks happening or why DMS fallbacks to _1 line per 1 line_.

You don't need to update your clients to the last version as it seems your problem is coming from server side.
Just update your server side with last pre-release version (**1375**)

Let me know if it's making a difference.

**Reply by MapInetum** [reply]
Hi @Mimetis.

The tests I have performed in my development environment seem satisfactory.

I still see a lot of deadlocks when the 'bulkupdate' procedure is executed on the server but at least the update operation/transaction is always retried with this bulk procedure.

As I say, this is in the development environment. This week I will deploy this version to the production environment and I will comment here the results.

Thanks as always!

Regards.

**Comment by MapInetum** [comment, marked as answer]
Hi @Mimetis.

Unfortunately the tests in the production environment are not being satisfactory.

It is observed that the execution of the bulk procedure on the server takes too long. Even more than with the previously used DMS version.

I have configured a PerBatchs synchronization and with a maximum of 1000 add/updates. 

![image](https://user-images.githubusercontent.com/79869950/215543666-6a45e61a-307a-40f7-920b-92c54e0ab9a1.png)


Now, I see that the problem may be in the instance of the server dbdd and how the lock is kept in the stored procedure.

Do you know any way to optimize this?

**Reply by Mimetis** [reply]
**18** min to inserts **1000 rows** ????

You have a problem somewhere ...

**Reply by Mimetis** [reply]
You are working with change tracking
What is your period retention ?
I guess you should look at the documentation and try to purge your tables:

https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/change-tracking-catalog-views-sys-change-tracking-tables?view=sql-server-ver16
https://learn.microsoft.com/en-us/sql/relational-databases/track-changes/cleanup-and-troubleshoot-change-tracking-sql-server?view=sql-server-ver16

**Reply by MapInetum** [reply]
Thirty days of retention period.

![image](https://user-images.githubusercontent.com/79869950/215851250-d6fd45fb-4e70-457d-91ed-75db70e43b9d.png)

**Reply by Mimetis** [reply]
I think it's too much in my opinion.
Maybe your bulkupdate is slow because your tracking table is too heavy (and scan takes too much time)
You should reduce the retention period and make a manual cleanup, and see if updates are still slow after

**Reply by MapInetum** [reply]
Hi @Mimetis.
Finally I manually clean and reduce de retention policy of change tracking and sync works fine.
Thanks for all.

**Reply by Mimetis** [reply]
did you gain a lot of performances since your clean up ?
I mean how percentage time did you gain?

**Reply by MapInetum** [reply]
My production configuration is:

- 8 MB batch size
- 2500 max batch lines
- Transactions per batch
- On server side, Read.Uncommited insolation and 7 days change tracking retention policy
- On client side, Read.Commited insolation and 30 days change tracking retention policy

I believe that with this configuration and after cleaning the change tracking the improvement has been more than 60-70%.

With these results, I have a stable and valid synchronization within the production infrastructure.

Thanks for everything.
